File: Solution/0015.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> tmp;
        
        sort(nums.begin(), nums.end()); // O(nlogn)
        int sum, num_idx1, num_idx2, num_idx3, idx2, idx3;
        // O(n^2)
        for(int i=0 ; i<nums.size()-2 ; ++i){
            // Skip duplicate elements
            if(i > 0 && nums[i] == nums[i-1]){
                continue;
            }
            num_idx1 = nums[i];
            
            // Initialization
            idx2 = i + 1;
            idx3 = nums.size() - 1;
            
            while(idx2 != idx3){
                num_idx2 = nums[idx2];
                num_idx3 = nums[idx3];
                sum = num_idx1 + num_idx2 + num_idx3;
                
                if(sum == 0){
                    tmp.push_back(num_idx1);
                    tmp.push_back(num_idx2);
                    tmp.push_back(num_idx3);
                    result.push_back(tmp);
                    tmp.clear();
                    idx2 += 1;
                    // Skip duplicate elements
                    while(nums[idx2] == nums[idx2 - 1] && idx2 < idx3)
                        idx2 += 1;
                }
                
                else if(sum > 0)
                    idx3 -= 1;
                else
                    idx2 += 1;
                
            }
            

        }
        
        return result;
    }
};

File: Solution/0015.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        sorted_arr = sorted(nums) # Sorted array O(nlogn)
        
        output_arr = []
        # O(n^2) worst case in while loop: 1+2+...+ (n-2)
        for idx1 in range(len(sorted_arr) - 2): 
            p1 = sorted_arr[idx1]
            if idx1 > 0 and p1 == sorted_arr[idx1-1]: # Dulicate cases
                continue
            
            idx2, idx3 = idx1+1, len(sorted_arr)-1
            while idx2 != idx3:
                p2, p3 = sorted_arr[idx2], sorted_arr[idx3]
                
                if p1+p2+p3 == 0:
                    if sorted_arr[idx2-1] == p2 and idx2-1 != idx1: # Dulicate cases
                        idx2 += 1
                    else:
                        output_arr.append([p1, p2, p3])
                        idx2 += 1
                elif p1+p2+p3 < 0:
                    idx2 += 1
                else:
                    idx3 -= 1

        return output_arr
    
    
